app/controllers/user_sessions_controller.rb

```diff
@@ -6,75 +6,132 @@ def new
+          redirect_to root_url, notice: "Already linked to another account!"
         end
-        @user_session = UserSession.new(username: params[:user_session][:username], 
-                                        password: params[:user_session][:password], 
-                                        remember_me: params[:user_session][:remember_me])
-        saved = @user_session.save do |result|
-          if result
-            # replace this with temporarily saving pwd in session,
-            # and automatically saving it in the user record after login is completed
-            if current_user.crypted_password.nil? # the user has not created a pwd in the new site
-              flash[:warning] = I18n.t('user_sessions_controller.create_password_for_new_site')
-              redirect_to '/profile/edit'
-            else
-              flash[:notice] = I18n.t('user_sessions_controller.logged_in')
-              if session[:openid_return_to] # for openid login, redirects back to openid auth process
-                return_to = session[:openid_return_to]
-                session[:openid_return_to] = nil
-                redirect_to return_to
-              elsif session[:return_to]
-                return_to = session[:return_to]
-                session[:return_to] = nil
-                redirect_to return_to
-              elsif params[:return_to]
-                redirect_to params[:return_to]
+      else # not signed in
+        if @identity&.user.present?
+          # The identity we found had a user associated with it so let's
+          # just log them in here
+          self.current_user = @identity.user
+          redirect_to root_url, notice: "Signed in!"
+        else #identity does not exist so we need to either create a user with identity OR link identity to existing user
+          if User.where(email: auth["info"]["email"]).empty?
+            #Create a new user as email provided is not present in PL database
+            user =  User.create_with_omniauth(auth)
+            @identity = UserTag.create_with_omniauth(auth, user.id)
+            key = user.generate_reset_key
+            # send key to user email
+            PasswordResetMailer.reset_notify(user, key).deliver_now unless user.nil? # respond the same to both successes and failures; security
+            redirect_to root_url, notice: "You have successfully signed in. Please change your password via a link sent to you via a mail"
+          else #email exists so link the identity with existing user and log in the user
+            user = User.where(email: auth["info"]["email"])
+            # If no identity was found, create a brand new one here
+            @identity = UserTag.create_with_omniauth(auth, user.ids.first)
+            # The identity is not associated with the current_user so lets
+            # associate the identity
+            @identity.save
+            #log in them
+            self.current_user = @identity.user
```

Comment:
here we current user is to be set as the user of identity. Any help appreciated.

Reply:
Where is an example of this being done under the previous system? Here? https://github.com/publiclab/plots2/blob/5597698ac59e4682ef3c2592aba2a10253292878/app/controllers/application_controller.rb#L72

Also this is becoming a very large function! I wonder if we should try to break out into some standardized way of a) identifying what scenario and b) passing minimum necessary parameters to a modular function that can do the necessary steps? Or some other way to help make this more readable? 

Thanks!

Reply:
This is to fetch current_user.
I need to assign current_user <- @identity.user.

Reply:
but within this, current_user is being assigned, no? i'm asking where in the existing code that is done, to look at for an example.

Reply:
Currently the method is able to fetch current user properly. So, I will say we can trim this function later on after this pr. 
I want ``def current_user=(user)`` function to assign current user.

Reply:
I marked that above.

Reply:
![image](https://user-images.githubusercontent.com/20972099/41245355-7777fa94-6dc5-11e8-8f74-e74ca0a89346.png)
We need to store the user in session. I think I am storing it at the wrong place.

Reply:
``session[:user_id] = user.id`` I think this line is error prone.

Reply:
OK, i looked through this and I believe it's these lines, which create and save a user_session:

https://github.com/publiclab/plots2/blob/e977a341512847997089942ddf04d2070be2015c/app/controllers/user_sessions_controller.rb#L31-L34

See how on this line, when fetching the session, it searches the same model? 

https://github.com/publiclab/plots2/blob/5597698ac59e4682ef3c2592aba2a10253292878/app/controllers/application_controller.rb#L69

Does that help? I think if you construct and save a session as in the first example, you should be good?